### ISSUE TITLE
Exported FMUs (CS and ME) systematic defect in handling "time" after simulation "restart"

### ISSUE THREAD

**Issue by janpeter19** [OpenModelica/OpenModelica#12561]
### Description
The subtle issue comes from exported FMUs, either CS or ME, and is a systematic defect in handling 'time' when simulations are restarted. With restarted I mean that you simulate up to a certain time, then (optionally) change some parameter, you update the models start values, and continue to simulate another period. The issue seems to occur just when 'time' is explicitly used.

### Steps to reproduce
The issue is illustrated well with the following model that use time in three different ways:
 - implicit use of time in a dynamical equations - no problem
 - explicit use of time just as a variable - initial problem after restart
 - explicit use of time using MSL.Sources.CombiTimeTable - initial problem that remains after restart

```modelica
    model Test_07
 	    import Modelica.Blocks.Sources; 
	    import Modelica.Blocks.Types;
	    parameter Real x_0 = 5;
	    Real x(start=x_0, fixed=true);
	    Real t;
	    Real u;
	    Sources.CombiTimeTable scheme(
	         smoothness=Types.Smoothness.ConstantSegments,
		 extrapolation = Types.Extrapolation.HoldLastPoint,
		 table=[0,0; 1.0,1; 2.0,2; 1003,3; 1004,4; 1005,5; 1006,6; 1007,7; 1008,8; 1009,9]);
    equation
           t = time;
	   u = scheme.y[1];
	   der(x) = -x;
    end Test_07;
```

--

In OMEdit the FMU is exported either as type CS or ME and use numerical solver CVODE.

The FMU is then run in a Python environment using PyFMI, or alternatively FMPy.
The same issue appears in both. I use here Ubuntu 22.04.

A screenshot is shown below using FMU of ME type.

![Screenshot 2024-06-12 at 11 18 51](https://github.com/OpenModelica/OpenModelica/assets/35634863/2b4ae967-13b2-4a8e-a52a-37d746bd3fb3)

### Version and OS
 - OpenModelica 1.23.0~dev.beta.1-1-g379f714
 - MSL 4.0.0
 - Linux Ubuntu 22.04 

### Additional context
Zip-file with the above mo-file,  a Jupyter notebook, and two PDFs of notebooks one for FMU-CS and one for FMU-ME are enclosed. At the end the PDFs show version numbers of key software used. 

[Issue - FMU handling time after restart.zip](https://github.com/user-attachments/files/15802334/Issue.-.FMU.handling.time.after.restart.zip)

**Comment by casella** [comment]
@arun3688 could you have a look at this when you have time? Thanks!

**Comment by janpeter19** [comment]
I just looked through release notes for 1.23.0.
Here was an issue #11855 that you fixed in January that may be closely related to my issue here.  But I am not that sure. Especially since I used 1.23.0-dev installed in May and I guess your fix propagated to 1.23-0-dev.beta.1-1-g379f714 which I use.

**Comment by arun3688** [comment]
@janpeter19 i am little busy with other works, I will try to look into this issue as early as possible thanks

**Comment by casella** [comment]
@janpeter19 in all cases please try with the latest nightly build first. Maybe the problem has been resolved already on the master branch.

**Comment by janpeter19** [comment]
@cassella,by far the easiest for me is download install the nightly build on VM Ubuntu that you have on your homepage. Tried that again. Problem is that the FMU-ME exported cannot even be loaded into Python environment. I took it up with Adrian in mid May and the problem was there also one year earlier. I have informed him and Martin, just now.

**Comment by janpeter19** [comment]
@casella, @arun3688  I have now installed a new VM with Ubuntu 22.04 and installed nightly build OMEdit 1.24.0~dev-185-g635218f - and the problem  described here is still there!

You can see it "live" in cell 14 in the notebook here https://colab.research.google.com/github/janpeter19/BPL_YEAST_AIR_Fedbatch/blob/main/BPL_YEAST_AIR_Fedbatch_colab.ipynb

**Comment by janpeter19** [comment]
@casella, @arun3688 I just checked that you have same problem for both numerical solver CVODE as for EXPLICIT EULER. Any progress?

**Comment by janpeter19** [comment]
@casella, @arun3688  I just checked with OM nightly build 1.25.0-dev-1-g7eb86c5 and still the same problem. So the task remains on your table :)

**Comment by casella** [comment]
@arun3688 please check at your earliest convenience

**Comment by janpeter19** [comment]
@casella, @arun3688  I just again tested with latest nighty buld 1.25-dev-14-g1a08d0b but still the same problem both me and cs types. Would you benefit from a proper  Linux PyFMI testbed so that you can run the test example conveniently yourself?

**Comment by casella** [comment]
@arun3688 please check at your earliest convenience

**Comment by arun3688** [comment]
@janpeter19 i did a quick testing on the model provided, i exported the fmu and tried to simulate with pyfmi, but it was showing me some issues related with encoding, So i did a quick test using `fmpy`  and the results looks correct for me, see the attached zip files, and also please try to test the fmu from fmpy and see whether you get the same problem what i can see the second time you  load the model and simulate the model with start_time= 2.5 and stop_time= 5.0 and what i see is there is no need to load the model second time and you can directly use the simulate_fmu in fmpy to start the simulation and from the plots we can see that the results are correct. see the attached zip file
[Test07.zip](https://github.com/user-attachments/files/17098068/Test07.zip)

My fmus have windows binaries, please use your fmu's when testing

**Comment by janpeter19** [comment]
@arun3688 I have tested your fmu_windows and fmpy-notebook in different ways. Essentially you suggest a "work-around" rather than addressing the core problem of the FMUs.  Not yet convinced it fully works for FMPy, but does not work for PyFMI even when I replace "re-loading" with "reset" when simulation continued. Here is what I have done. 

1. Your FMU run in my pyfmi-script enclosed in earlier post in this thread reproduce the problem. Thus I do not have your problem with pyfmi (but I get a warning about could not find GLIMDA first time you simulate but not serious). I suspect something wrong in you pyfmi-installation in Windows. Note that I setup an environment with Miniconda for pyfmi. Further, I have not anything OpenModelica related in my Windows (VBox).
 
2. Your FMU run in your enclosed FMPy-script works fine and I reproduce your results in Windows.

3. My Linux FMU run in your notebook FMPy in Linux (VBox) do also reproduce your results. My FMU was made recently with OM 1.25-dev. (Somehow notebook not possible to print to PDF).

I agree that re-loading the FMU should not be necessary, but instead I think some "reset" mechanism of the FMU should be used. 

4. In my new enclosed PyFMI-windows script using reset instead of loading, I get correct results with FMU generated with JModelica but your FMU from OpenModelica does not work. This fact should worry you too!

It would perhaps be helpful to see a complete FMPy-script of yours that generate my second figure. Surprising to me that you do not use some "reset" functionality, but I am less familiar with using FMPy. 

Here is a post on FMPy and use of reset that might be useful. But I do not fully understand it. Perhaps you can make use of it?
https://github.com/CATIA-Systems/FMPy/issues/372

Even if we may make my example work with FMPy I think it would be a pity not to be able to use it in PyFMI. So I tend to think you should address the core problem with FMU generated from OpenModelica. What do you think? 

Enclosed PDFs of notebooks i have run and described above, except Linux notebooks that recently changed and do not know how to print to PDF.

[Test07_continued.zip](https://github.com/user-attachments/files/17114810/Test07_continued.zip)

**Comment by arun3688** [comment]
@janpeter19 I need to fix `pyfmi` in my laptop to reproduce the issue and hopefully see what is going wrong, It would be great to see if you can generate different result files after the model is reset and plot the graph, i am not sure whether pyfmi has that support, but on  a quick note 

```
# Instead just do reset
model.reset()
for key in stateDict.keys():
 model.set(key+'_0', stateDict[key])
# Simulation 2
sim_res = model.simulate(start_time=2.5, final_time=5) 
```
can you just store the result in another variable instead of sim_res and see what happens, for example

`sim_res_new = model.simulate(start_time=2.5, final_time=5) 
`
and then try to plot the graph with `sim_res_new` and see if it produces the same behaviour

**Comment by janpeter19** [comment]
I just like to add that the warning  in PyFMI "Could not find GLIMDA" only comes when running FMU from OpenModelica not from JModelica.  You see that in notebook 4a in the previous post. So I think the warning has something to do with the FMU generation in OpenModelicar, rather than your PyFMI installation.

**Comment by janpeter19** [comment]
I updated to PyFMI 2.14.0 the other day and you see what Python version I have. Recall vaguely that in the summer hade some problems with some update 2.13.1 or so, problems related to Assimulo version. I am not so good att interpreting the error text you have though. 

You also wanted me to just run the second simulation storing in sim_res_new[] but I do get the same results. Notebook enclosed. 

[4c_Test_07c_pyfmi_windows_om.pdf](https://github.com/user-attachments/files/17115465/4c_Test_07c_pyfmi_windows_om.pdf)

**Comment by arun3688** [comment]
@janpeter19 thanks for testing the suggestion, I can see something is going wrong, so I need to debug it, i will get back to you once i found the problem, i will try to update the pyfmi with the latest version, may be i will compile it from source because pyfmi install via pip package is too old it seems it has pyfmi version 2.5 which seems to be very old.

**Comment by casella** [comment]
> pyfmi install via pip package is too old it seems it has pyfmi version 2.5 which seems to be very old.

I think we already noticed this when we started testing FMUs seriously for Buildings. Maybe we opened a ticket on their issue tracker about that?

**Comment by bilderbuchi** [comment]
> > pyfmi install via pip package is too old it seems it has pyfmi version 2.5 which seems to be very old.
> 
> I think we already noticed this when we started testing FMUs seriously for Buildings. Maybe we opened a ticket on their issue tracker about that?

this is https://github.com/modelon-community/PyFMI/issues/210 and https://github.com/modelon-community/PyFMI/issues/242
I agree regarding using miniconda, instead, that has 2.14.

**Comment by janpeter19** [comment]
I tried myself to make fmpy script complete to reproduce my original figure 2. Then I find that FMU from OpenModelica actually do not work. Here is a zero at the new start time that becomes wrong in variable "time" as well as from output from CombiTimeTable variable u. The FMU from JModelica works, as expected. Perhaps this observation can help you in your deeper troubleshooting?

![Test07_windows_om_fmpy](https://github.com/user-attachments/assets/f2efc209-8fe7-48fe-baf5-50b95de627d3)

![Test07_windows_jm_fmpy](https://github.com/user-attachments/assets/2c53d844-42e8-489a-a4d1-5fc19019d3f8)

**Comment by arun3688** [comment]
@janpeter19 thanks for your tests in fmpy, can you share the notebook which contains the fmpy

**Comment by janpeter19** [comment]
@arun3688 enclosed you find a zip-file with the fmpy-notebook you see in the post above as well as the two FMU used. One I did with JModelica and the others is yours from OpenMOdelica but renamed for clarity.

[Test07_continued3.zip](https://github.com/user-attachments/files/17143446/Test07_continued3.zip)

I have no experience of troubleshooting at your level. But as an outsider I think you should trace how the parameter "start_time" is propagated. I suspect a very simple error that a default value 0 is not replaced with "start_time". Earlier (2024-09-02) I reported that the problem was the same for CVODE and Explicit Euler.  Note also that (at least in Linux) here is a difference between FMU and CS and ME type. The difference is the output u (from CombiTimeTable) where for CS you only have  spike downward  to 0 at time 2.5 and the you continue to have a correct value 2, while for ME the value remain at 0 after 2.5. Test yourself in Windows by making two different FMUs (yours is a combo of both ME and CS and behave in this case as CS). I hope these observations can help you zoom in on where the mistake is in the code.

**Comment by arun3688** [comment]
@janpeter19 I can reproduce the issue now, I will debug more and see what is causing the issue

**Comment by arun3688** [comment]
@janpeter19 the issue is fixed with this commit 4d935ff2761f54fccd68d46f01137857af3e8525, Please test it from nightly builds today (28-9-2024). i exported the fmu and tested with Fmpy, the output looks correct see the attached zip with the new exported fmu 
[Test07_Fixed.zip](https://github.com/user-attachments/files/17171487/Test07_Fixed.zip)

**Comment by janpeter19** [comment]
@arun3688 your results with your enclosed windows FMU of type MECS, I have confirmed both with FMPy and PyFMI. The PyFMI files are enclosed with your FMU for completeness. 

[Test07_Fixed_pyfmi.zip](https://github.com/user-attachments/files/17173657/Test07_Fixed_pyfmi.zip)

The nightly build in Linux did not manage to produce an FMU that is corrected though. In Linux the nightly build is called '...g149ccc2' while in Windows  you have "...ge813f88fed'.  The commit history of Linux nightly build show that your change should be there. But, actually there is a red cross and seems that commit was not taken?

![Screenshot 2024-09-28 at 10 07 07](https://github.com/user-attachments/assets/50cdb85e-4a98-42f9-a887-ead0532d9185)

**Comment by arun3688** [comment]
@janpeter19 You are correct, the nightly builds in linux did not take my commit, but the windows nightly builds was successfully built, But i hope with the fmu's i provided it works as expected, You can try to test it from the nightly builds on monday (30-09-2024), Hopefully the linux build is fixed

**Comment by janpeter19** [comment]
@arun3688 @casella  Your commit Sep 28 has some problem on the Linux side and  if its not your "bag" then perhaps someone else should be consulted? Further, your faulty commit block another commit done Sep 30 on the Linux side. Likely a small and silly problem, or?

**Comment by arun3688** [comment]
@janpeter19 it is not faulty commit, but somehow the nightly builds on the linux is failing or may be you are still running with old openmodelica version, it would be good if you can uninstall your older version and then install the new nightly builds hopefully that should solve the problem

**Comment by janpeter19** [comment]
@arun3688 The public list of Commits I interpret as the are problems with yours, and also affect adeas31 commit yesterday. Has nothing to do with my installation. See the chart below. But I am rather new to OpenModelica update logs and may misunderstand something...

![Screenshot 2024-10-01 at 10 46 28](https://github.com/user-attachments/assets/50c2f4cc-b8e1-4ef5-8e2a-a2cec8fcb22c)

**Comment by arun3688** [comment]
@janpeter19 that issue is not related with this commit,  the issue is related with the server, so it has nothing to do with nightly builds, the commits are usually tested in linux and we cannot merge the commit, until all the tests are passed, after we merge then commit is run on the master branch once again and it failed with some server problem, you can also see some commits which fails after we merge to master

**Comment by casella** [comment]
@janpeter19 for some reason the Linux [nightly build jobs](https://test.openmodelica.org/jenkins/job/LINUX_BUILDS/) seem to be running fine but the apt server is not being updated for some reason. We're investigating that right now.

**Comment by janpeter19** [comment]
@arun3688 Strange to me. I got into the website

https://openmodelica.org/download/download-linux/

In Nightly Build I click on commit history.  I see your commit and a "red cross". 

Is there a Linux build that I can use? What name has it? Link? Or I should just wait. I am off now for a few hours anyway.

**Comment by arun3688** [comment]
@janpeter19 it seems there is some problem with apt server as casella said, So you have to wait until this is solved

**Comment by casella** [comment]
@janpeter19 I'm not sure what those red crosses mean, but I don't think they are relevant. Pull requests are merged in only if they pass all the tests.

**Comment by casella** [comment]
@janpeter19, job [#2822](https://test.openmodelica.org/jenkins/job/LINUX_BUILDS/2822/) is building now. If it succeeds, you should be able to run apt update/upgrade and get the new nightly build.

**Comment by janpeter19** [comment]
@arun3688 @casella Now it works  using nightly buid 1.25-0-dev-51-ge672d09.

Thank you!

**Comment by janpeter19** [comment]
@arun3688 @casella I have gone through about 15 test cases and I actually get problem in OMEdit to export the FMU, although model check as well as simulation in OMEdit works. The identical code JModelica compile to a working FMU. OMEdit give instead a complaint about cMake, see below. How should I understand this message? What information more to get out? 

![Screenshot 2024-10-07 at 21 11 53](https://github.com/user-attachments/assets/9814534c-aa4a-4b7b-ae19-50fa64e35938)

This application is more complex than my other test cases and you see the process diagram below. 

![BPL_IEC_process_diagram_om](https://github.com/user-attachments/assets/dc7d697a-78dd-4313-8708-95c2838318da)

I have not been able yet to check the model on a somewhat older version of OMEdit to see that it works there.

If this post does not really belong here, let me know, and I take it away.

**Comment by arun3688** [comment]
@janpeter19 Can you able to export the fmu's on other simple models ? Based on the error message i see that CMakeLists.txt is not generated which means the model failed in the translation stage, if you can provide the model i can debug it

**Comment by janpeter19** [comment]
All other models I have no problem with and same nightly build version.
The model is complicated to share, not open source. How should I understand the error text?

**Comment by arun3688** [comment]
> All other models I have no problem with and same nightly build version.

that is good, because the new fmu export is not tested with linux and it is good that it works

> The model is complicated to share, not open source. How should I understand the error text?
the `CmakeList.txt` is not generated in the sources directory which means something might have happened during translating model, Can you run the below mos script and see. 

### a.mos
```
loadModel(Modelica);getErrorString();
loadFile("path to library.mo"); getErrorString();
translateModelFMU("modelName"); getErrorString();
```
change the `filename` and `modelName` according to your path and model name and then run from the terminal

`>> omc a.mos`

**Comment by arun3688** [comment]
@janpeter19 It is really difficult to tell what is the problem with the above screenshot, with the screenshot i can confirm that cmake build is failing and it is failing because there is no `CMakeLists.txt` generated. you have to run the following 

a.mos
```
loadModel(Modelica);getErrorString();
loadFile("V:/GitHub/BPL_GUI/BPL/package.mo"); getErrorString();
translateModelFMU("BPL_IEC.Column_system"); getErrorString();
```
please save the file as a.mos and run the script

`>>> omc a.mos`

**Comment by arun3688** [comment]
@janpeter19 in my opinion this should be opened as a separate issue and it has nothing to do with fmu_export

**Comment by janpeter19** [comment]
@arun3688 well, likely you right. But at least in "theory" it could be a side-effect of the latest fix to make export FMU generate a truly functioning FMU. The way to make sure it is to reproduce the error in an older version of OMEdit, say 1.24.0-beta or so. Usually I have one VBox for nightly build and one VBox for stable versions of OMEdit. Yesterday, found my VBox for the stable version to not work any more, and I have to rebuild and manually install OMEdit 1.24... stable. This takes for me we couple of hours to do. The ready-made VBoxes you can download, here the FMUs does not work for some other reasons. I have made the LInkoping team aware a few times before, and I do not think it is fixed yet. One of these days I take the time to make a new VBox. Provided my problem can be reproduced there, then I make a new thread.  Ok?

**Comment by janpeter19** [comment]
@arun3688 The issue is partly solved and has nothing to do with this post. So I agree with you and deleted the last few posts of mine. I suggest you do the same. I may come back on the phenomena under another heading. It has to do with differences between "within" and "import" and differences JModelica and OpenModelica. Not sure what is right, but I have a procedure of coding that now works for both platforms, although less convenient.

**Comment by arun3688** [comment]
@janpeter19 I have deleted my comments and we can close this issue
